### PR TITLE
Feature: allow setting up custom environment when creating windows and/or panes

### PR DESCRIPTION
**Pull request by zappolowski**
This is required for implementing https://github.com/tmux-python/tmuxp/issues/832.

### CODE REVIEW

**Comment by tony** [comment]
@zappolowski FYI on timeline, I may not be able to give this my undivided attention until the evening (weekends latest)

File: src/libtmux/session.py [hunk 1]
```diff
@@ -259,6 +260,10 @@ def new_window(
             % (self.id, window_index),
         )
 
+        if environment:
+            for k, v in environment.items():
+                window_args += ("-e%s=%s" % (k, v),)
```
Comment (tony) [inline src/libtmux/session.py:265]: tmux 3.0 compatible: https://github.com/tmux/tmux/blob/42358cc/CHANGES#L823

File: src/libtmux/window.py [hunk 2]
```diff
@@ -520,6 +521,10 @@ def split_window(
         if not attach:
             tmux_args += ("-d",)
 
+        if environment:
+            for k, v in environment.items():
+                tmux_args += ("-e%s=%s" % (k, v),)
```
Comment (tony) [inline src/libtmux/window.py:526]: tmux 3.0+ compatible (tmux 3.0 compatible: https://github.com/tmux/tmux/blob/42358cc/CHANGES#L823)

Comment (tony) [reply src/libtmux/window.py:526]: We could add a version check (and a warning) for tmux 3.0+ (same with the other case + test)

```python
from libtmux.common import has_gte_version

if has_gte_version("3.0"):
```

File: src/libtmux/session.py (hunk 1 above)
Comment (zappolowski) [reply src/libtmux/session.py:265]: Will take a look.

Comment (zappolowski) [reply src/libtmux/session.py]: You can take another look. If you're okay with the changes I'd fixup the commits and push them again.

File: src/libtmux/session.py [hunk 3]
```diff
@@ -259,6 +261,15 @@ def new_window(
             % (self.id, window_index),
         )
 
+        if environment:
+            if has_gte_version("3.0"):
+                for k, v in environment.items():
+                    window_args += ("-e%s=%s" % (k, v),)
```
Comment (tony) [inline src/libtmux/session.py:267]: Bikeshed / fine to resolve:

This is what pyupgrade gives me

```suggestion
                    window_args += ("-e{}={}".format(k, v),)
```

You could also try f-strings

```suggestion
                    window_args += (f"-e{k}={v}",)
```

I am not sure if one is better than the other. 

You can change to f-string or keep it as-is and resolve

File: tests/test_window.py [hunk 4]
```diff
@@ -310,3 +312,45 @@ def test_empty_window_name(session: Session) -> None:
...
+    session: Session,
+    caplog: pytest.LogCaptureFixture,
+) -> None:
+    env = shutil.which("env")
+    assert env is not None, "Cannot find usable `env` in Path."
+
+    window = session.new_window(window_name="split_window_with_environment")
+    window.split_window(
+        shell=f"{env} PS1='$ ' sh",
+        environment={"ENV_VAR": "pane"},
```
Comment (tony) [inline tests/test_window.py:351]: Are you familiar with [pytest.mark.parametrize](https://docs.pytest.org/en/6.2.x/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions)? Any objection to adding some examples with more variations of env variables?

Is this going to assert the variable is set in some way?

**Review by tony — APPROVED** [review]
LGTM

Fine to squash the commits as you see fit

File: src/libtmux/session.py (hunk 1 above)
Comment (tony) [reply src/libtmux/session.py]: Looks good to me (to fix up)

File: src/libtmux/window.py (hunk 2 above)
Comment (tony) [reply src/libtmux/window.py]: Looks good!

File: tests/test_window.py (hunk 4 above)
Comment (zappolowski) [reply tests/test_window.py:351]: I'm not familiar with this one, but can take a look (probably tomorrow earliest). Any specific idea of what kind of different environment variables might be beneficial to test (in this test-case it won't make a difference as the test checks whether the correct warning is shown to the user when an old version of tmux is used)?

> Is this going to assert the variable is set in some way?

For this test it's not asserted as it won't end up anywhere. It's just to use the same setup as the test case which actually sets up the environment and for that test the assertion happens in line 334.

**Comment by tony** [comment]
@zappolowski Game plan

I will merge #455 and make a patch release for 0.15.9

we can then merge this (i will rebase) and do final check, if it looks good, will release 0.16.0

**Comment by tony** [comment]
@zappolowski 

v0.15.10: [GitHub Release](https://github.com/tmux-python/libtmux/releases/tag/v0.15.10), [docs site](https://libtmux.git-pull.com/history.html#libtmux-0-15-10-2022-11-05), [change file](https://github.com/tmux-python/libtmux/blob/v0.15.10/CHANGES#libtmux-01510-2022-11-05), [PyPI](https://pypi.org/project/libtmux/0.15.10/)

File: tests/test_window.py (hunk 4 above)
Comment (tony) [reply tests/test_window.py:351]: > For this test it's not asserted as it won't end up anywhere

Good point.

This whole conversation was meant to be in `test_split_window_with_environment` - where for some reason I didn't see the `assert. My mistake.

File: tests/test_window.py [hunk 5]
```diff
@@ -310,3 +312,45 @@ def test_empty_window_name(session: Session) -> None:
         "#{==:#{session_name}," + session.name + "}",
     )
     assert "''" in cmd.stdout
+
+
+@pytest.mark.skipif(
+    has_lt_version("3.0"),
+    reason="needs -e flag for split-window which was introduced in 3.0",
+)
+def test_split_window_with_environment(session: Session) -> None:
```
Comment (tony) [inline tests/test_window.py:321]: > I'm not familiar with this one, but can take a look (probably tomorrow earliest

Fine if I add, or would you like to? Up to you

> Any specific idea of what kind of different environment variables might be beneficial to test

- One variable (as you have now): `{"ENV_VAR": "pane"}`
- Two or more: `{"ENV_VAR": "pane", "ENV_VAR_2": "another variable"}`

File: tests/test_window.py (hunk 4 above)
Comment (tony) [reply tests/test_window.py:351]: Moved the rest of the conversation to the appropriate test https://github.com/tmux-python/libtmux/pull/453/files#r1014695165

File: tests/test_window.py (hunk 5 above)
Comment (zappolowski) [reply tests/test_window.py]: I've extended the tests for `Session.new_window` and `Window.split_window`.

Comment (tony) [reply tests/test_window.py]: @zappolowski LGTM, you can squash

Then I will make a release for libtmux and update tmuxp master to point to this

**Comment by tony** [comment]
@zappolowski I didn't get a notification of the push!  Merging and publishing now

**Comment by tony** [comment]
@zappolowski 

- [v0.16.0a0](https://github.com/tmux-python/libtmux/tree/v0.16.0a0) ([PyPI](https://pypi.org/project/libtmux/0.16.0a0/)) is now live 
- tmuxp has libtmux v0.16.0a0 (with this PR) as of [5c222fa](https://github.com/tmux-python/tmuxp/commit/5c222facbfc95e433bb6fb07d0774d1ac455bf46)

The floor is yours to address https://github.com/tmux-python/tmuxp/issues/832#issuecomment-1298879871

**Pull request merged**